plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/AbstractIcebergTableOperations.java

```diff
@@ -229,6 +233,28 @@ protected void refreshFromMetadataLocation(String newLocation)
         shouldRefresh = false;
     }
 
+    private TableMetadata readTableMetadata(String metadataLocation)
+    {
+        try {
+            return TableMetadataParser.read(fileIo, io().newInputFile(metadataLocation));
+        }
+        catch (Exception e) {
+            if (getCausalChain(e).stream().anyMatch(AbstractIcebergTableOperations::isNotFoundError)) {
+                throw new TrinoException(ICEBERG_METADATA_NOT_FOUND, e);
+            }
+            throw e;
+        }
+    }
+
+    private static boolean isNotFoundError(Throwable throwable)
+    {
+        if (throwable instanceof org.apache.iceberg.exceptions.NotFoundException) { // qualified name, as this is NOT the io.trino.spi.connector.NotFoundException
+            return true;
+        }
+        return throwable instanceof UnrecoverableS3OperationException &&
```

Comment:
> We shouldn't bring in `AbstractIcebergTableOperations` references to `s3`

I agree, it would be better not to have FS-specific exception handling here. cc @losipiuk @electrum for ideas 

> We shouldn't bring in AbstractIcebergTableOperations references to s3 because this layer should remain abstract / free from implementation details.
We can however override the method isNotFoundError for certain implementations of this abstract class.

@findinpath this won't help. The subclasses are specific to the metastore they work with, not for the filesystem.

Reply:
Actually, I hesitated to put file system specific handling in this class when writing the code, but I didn't find the simple solution. Any ideas are welcome.

Reply:
Changed to `TrinoException(HIVE_FILE_NOT_FOUND)` instead of exposing the error code via `AmazonS3Exception`.